builder: Redo git submodules handling

bochecha opened the pull request:
`flatpak-builder` was mixing the path and the name of a submodule, which would lead to download failures when they were different.

Fixes #216

bochecha commented:
I'm getting failures in the unit tests, but I'm not sure that's related to this change. I don't really understand why they fail though. :confused:

bochecha commented:
I get the same test failures on master, so it doesn't seem to be coming from my change.

bochecha commented:
Just found their was an `uncrustify.sh` script in the source tree, so I ran it and amended my commit with the stuff it fixed in my change.

alexlarsson commented on builder/builder-source-git.c:
```diff
@@ -279,70 +279,54 @@ git_mirror_submodules (const char     *repo_location,
+          name = g_strndup (submodule + 11, len - 12);
+          path = g_key_file_get_string (key_file, submodule, "path", error);
+          url = g_key_file_get_string (key_file, submodule, "url", error);
 
```
You're reusing "error" here, which may have been set in the call on the previous line. That would cause it to break.

alexlarsson commented on builder/builder-source-git.c:
```diff
@@ -279,70 +279,54 @@ git_mirror_submodules (const char     *repo_location,
 
-          if (!git_mirror_repo (url, update, words[0], context, error))
+          if (!git (mirror_dir, &ls_tree, error, "ls-tree", revision, path, NULL))
+            continue;
```
Same here. You either need to print and clear a local error, or fail the whole operation (return false) and modify "error".

alexlarsson commented on builder/builder-source-git.c:
```diff
@@ -279,70 +279,54 @@ git_mirror_submodules (const char     *repo_location,
+          if (!git (mirror_dir, &ls_tree, error, "ls-tree", revision, path, NULL))
+            continue;
+
+          words = g_strsplit_set (ls_tree, " \t", 4);
```
Maybe we can add some verification here on the ls_tree output. For instance, make sure its only one line (i.e. it would not if the path was a directory), and at least one line (happens if the path didn't exist). I guess you want to split on \n first. Also, words[1] should be "commit".

alexlarsson commented on builder/builder-source-git.c:
```diff
@@ -417,75 +401,80 @@ builder_source_git_download (BuilderSource  *source,
 
-  if (submodule_status)
+  if (git (checkout_dir, &submodule_data, NULL, "show", gitmodules, NULL))
     {
```
How does this handle the case where there is not submodules? I.e. if git returns a nonzero exist status?

bochecha replied:
In that case the `git` function returns `FALSE`, and we do not enter the subsequent block.

However, running that ` git show` command will print an error message if there aren't any submodules:

```
fatal: Path '.gitmodules' does not exist in '5367783d35897e2d2a832ba7805ec9f050997a37'
```

That's a bit ugly, but it is completely harmless. However it might confuse people.

So I'm not sure what to do about this.

alexlarsson commented on builder/builder-source-git.c:
```diff
@@ -417,75 +401,80 @@ builder_source_git_download (BuilderSource  *source,
-          g_print ("processing submodule %s\n", words[1]);
+          path = g_key_file_get_string (key_file, submodule, "path", error);
+          url = g_key_file_get_string (key_file, submodule, "url", error);
 
```
same error issue here

alexlarsson commented on builder/builder-source-git.c:
```diff
@@ -417,75 +401,80 @@ builder_source_git_download (BuilderSource  *source,
 
-          mirror_dir = git_get_mirror_dir (child_url, context);
+          if (!git (checkout_dir, &ls_tree, error, "ls-tree", revision, path, NULL))
+            continue;
```
same error issue here

bochecha commented:
I **think** I have now fixed the issues with the way I was handling errors. I might have misunderstood how to properly fix those, though, as C is not my mother tongue. :smiley: 

I have also added some checks you requested to verify that the `ls-tree` is a valid `gitlink`:
* only one line
* the second word (`words[1]`) is `"commit"`

The tests still fail, just like they do on master so I assume this change doesn't break them.

Let me know if there is anything else to fix in this pull-request.

alexlarsson commented:
pushed this (with minor merge issue) to master

alexlarsson commented:
I think this broke gedit:
http://sdkbuilder1.gnome.org/logs/build-2016-08-23-103003/build-gnome-apps-nightly-master-org.gnome.gedit-x86_64.txt

bochecha commented:
I don't think it's related: https://git.gnome.org/browse/gedit/tree/.gitmodules#n3

bochecha commented:
You were right, I did break the gedit build, see #263 for the fix
